sycl/source/detail/plugin.hpp

```diff
@@ -163,7 +164,13 @@ class plugin {
 #endif
     RT::PiResult R;
     if (pi::trace(pi::TraceLevel::PI_TRACE_CALLS)) {
-      std::lock_guard<std::mutex> Guard(*TracingMutex);
+      // MShutdownModeEnabled signals if process is terminating now. It covers
+      // normal case and undefined behavior when it happens by initiative from
+      // underlying modules (calling exit(1)). In case of unexpected exit(1)
+      // call sync primitive can be still locked.
+      auto Guard = MShutdownModeEnabled
+                       ? std::unique_lock<SpinLock>()
+                       : std::unique_lock<SpinLock>(*MTracingSLock);
```

Comment:
So, you avoid locking if tracing called at shutdown, right? Can you explain what is the UB being addressed here? Also, is this really the only lock that is problematic in case of process termination?

Reply:
sure, you understanding is correct. If anyone inside pi call calls exit(1) (e.g. assert failure) and traces are enabled we will have a hang because we have:
- mutex.lock
- pi call that calls exit() and never returns
- unreachable mutex.unlock;
Exit() causes normal process termination and provides opportunity to release resources. In this case we do shutdown and call piTearDown that is also wrapped by the same mutex.
In this case right now RT just hangs.

Reply:
For the problematic places:
the situation with exit in backends may affect two places that I know:
1) PI calls tracing - environment variable is documented; addressed by this PR
2) ze calls tracing - ZE_SERIALIZE=1, not documented, fully internal, not addressed. https://github.com/intel/llvm/blob/sycl/sycl/plugins/level_zero/pi_level_zero.cpp#L51

Reply:
double checked - ze calls tracing is not affected, piTearDown doesn't call any

Reply:
> If anyone inside pi call calls exit(1) 

Which code does that? Maybe we should disallow exit() from plugins? It should only return error codes to SYCL RT or terminate() in case of fatal error (and no shutdown() in this case).

Reply:
yes, you are right. We requested this too.

Reply:
> unreachable mutex.unlock

It's still unreachable here. So, we just want to avoid attempting to lock already locked mutex (spinlock) and lock a new mutex instead, don't we?

Reply:
I still think that this is a workaround for a bigger problem that plugins leave things in corrupted state (causing UB). Can we instead fix the plugins (to not exit, or only exit after all resources are released)?

Reply:
> I still think that this is a workaround for a bigger problem that plugins leave things in corrupted state (causing UB)

Right, but it's not plug-ins being guilty but backends.
It's requested from backend guys to fix the behaviour so, that they don't terminate thread/process. @dm-vodopyanov , please, correct me if I'm wrong.

Reply:
Maybe we should just turn the mutex around PI call to be "timed_mutex"? Also, holding the mutex around some PI calls (e.g. piProgramBuild) may be too expensive and freeze other threads for no good reasons. The "timed_mutex" will only hold  it for, say 1 sec. Yes, some lengthy calls will have it's output mixed with other threads output. That would be addressed in long-term by using XPTI for tracing.

Reply:
Why is there a need for mutex at all?
Is it only due multiline output for PI tracing?
Is it feasible to switch to single-line output like e.g. strace?
Any implications for such a change?

@alexbatashev , do you have anything to advise here?

Reply:
The mutex is to make trace readable, i.e. such that output from multiple threads is not mixed. Remember, some prints are **before** PI call, some **after**, and some other are from **within** the PI call.

Reply:
@s-kanaev I'm not sure if we need to change anything at all. The original hang was caused by a bug in the underlying runtime. This workaround is for that specific case, there's no guarantee that it won't fire in another case. XPTI tracing apparently is not affected by this issue, and it is almost ready to be merged: https://github.com/intel/llvm/pull/5389, so, we should probably focus on improving that tool, rather than putting some questionable w/as all over runtime. If we still want to fix SYCL_PI_TRACE, we @smaslov-intel has a good advice - timed mutex with long enough timeout. That'll be less of a hack than what's proposed by this PR.

Reply:
https://en.cppreference.com/w/cpp/thread/timed_mutex/~timed_mutex

timed mutex has the same UB conditions as std::mutex. Not sure if it would be safe enough to use it.

The behavior is undefined if the mutex is owned by any thread or if any thread terminates while holding any ownership of the mutex.